server/game-service/main.go

```diff
@@ -339,89 +244,182 @@ func handleConnection(w http.ResponseWriter, r *http.Request) {
-				CardIndex int    `json:"cardIndex"`
-			}
-			if err := json.Unmarshal(message, &payload); err == nil {
-				switch payload.Action {
-				case "move":
-					handleMove(gameID, playerIdx, payload.Matched)
-				case "quit":
-					handleQuit(gameID, playerIdx)
-				case "flip":
-					handleFlip(gameID, playerIdx, payload.CardIndex)
-				}
+// Function to handle player disconnection
+func handleDisconnection(gameID string, playerIdx int) {
+	gamesMux.Lock()
+	game, exists := games[gameID]
+	gamesMux.Unlock()
+	if !exists {
+		return
+	}
+
+	game.Mutex.Lock()
+	defer game.Mutex.Unlock()
+
+	if game.GameStatus == "match_end" {
+		return
+	}
+
+	fmt.Printf("[WAITING] Player %d disconnected | Waiting 10s before declaring other player winner | Game ID: %s\n",
+		playerIdx, gameID)
+
+	go handlePlayerTimeout(game, playerIdx, gameID)
+}
+
+// Function to handle the player timeout
+func handlePlayerTimeout(game *Game, playerIdx int, gameID string) {
+	time.Sleep(10 * time.Second)
+	game.Mutex.Lock()
+	defer game.Mutex.Unlock()
+
+	if game.Players[playerIdx] == nil {
+		game.GameStatus = "match_end"
+		winner := 1 - playerIdx
+		game.Winner = winner // Set the winner
+		fmt.Printf("[WIN] Player %d wins by default | Game ID: %s\n", winner, gameID)
+		// Send game state to remaining players
+		sendGameState(game, gameID)
+	}
+}
+
+// Function to listen for messages from client
+func listenForMessages(conn *websocket.Conn, game *Game, playerIdx int, username, gameID string) {
+	defer conn.Close()
+	for {
+		_, message, err := conn.ReadMessage()
+		if err != nil {
+			fmt.Printf("[DISCONNECTED] Player %d (%s) closed connection | Game ID: %s\n", playerIdx, username, gameID)
+			game.Mutex.Lock()
+			game.Players[playerIdx] = nil
+			game.Mutex.Unlock()
+			handleDisconnection(gameID, playerIdx)
+			return
+		}
+
+		var payload struct {
+			Action    string `json:"action"`
+			Matched   bool   `json:"matched"`
+			CardIndex int    `json:"cardIndex"`
+		}
+		if err := json.Unmarshal(message, &payload); err == nil {
+			switch payload.Action {
+			case "move":
+				handleMove(gameID, playerIdx, payload.Matched)
+			case "quit":
+				handleQuit(gameID, playerIdx)
+			case "flip":
+				handleFlip(gameID, playerIdx, payload.CardIndex)
 			}
 		}
-	}()
+	}
 }
 
-// This function handles player quitting
+// Function to handle a player's move
+func handleMove(gameID string, playerIndex int, matched bool) {
+	gamesMux.Lock()
+	game, exists := games[gameID]
+	gamesMux.Unlock()
+	if !exists {
+		fmt.Println("[ERROR] Game not found:", gameID)
+		return
+	}
+
+	game.Mutex.Lock()
+	defer game.Mutex.Unlock()
+
+	if game.GameStatus == "match_end" {
+		fmt.Println("[INFO] Move ignored, game is over")
+		return
+	}
+
+	if game.CurrentPlayer != playerIndex {
+		fmt.Printf("[WARNING] Player %d tried to move out of turn in Game ID: %s\n", playerIndex, gameID)
+		return
+	}
+
+	if matched {
+		handleMatch(game, playerIndex, -1)
+	} else {
+		handleTurnSwitch(game, playerIndex)
+	}
```

Comment:
keep the guard clauses, you already are using them why suddenly switch?


```suggestion
    if matched{
        handleMatch(game, playerIndex, -1)
        return
    }

    handleTurnSwitch(game, playerIndex)
```

Reply:
+I have implemented this already